packages/xstate-store/src/atom.ts

```diff
@@ -1,85 +1,212 @@
+  initialValue: T,
+  options?: AtomOptions<T>
+): Atom<T>;
+export function createAtom<T>(
+  valueOrFn: T | ((read: <U>(atom: Readable<U>) => U) => T),
+  options?: AtomOptions<T>
 ): Atom<T> | ReadonlyAtom<T> {
-  const current = { value: undefined as T };
-  let observers: Set<Observer<T>> | undefined;
-
-  // Handle computed case
   if (typeof valueOrFn === 'function') {
-    const subs = new Map<AnyAtom, Subscription>();
-    let observedAtoms = new Set<AnyAtom>();
-
-    const getValue = valueOrFn as (read: <U>(atom: Atom<U>) => U) => T;
-    const read = (atom: AnyAtom) => {
-      observedAtoms.add(atom);
-      const val = atom.get();
-      if (subs.has(atom)) {
-        return val;
+    return computed(valueOrFn as () => T, options);
+  }
+
+  // Create plain object atom
+  const atom: Atom<T> = {
+    currentValue: valueOrFn,
+
+    // Dependency fields
+    subs: undefined,
+    subsTail: undefined,
+
+    get(): T {
+      if (activeSub !== undefined) {
+        link(this, activeSub);
       }
-      const sub = atom.subscribe(recompute);
-      subs.set(atom, sub);
-      return val;
-    };
-
-    function recompute() {
-      observedAtoms = new Set();
-      const newValue = getValue(read);
-
-      // Cleanup any atoms that are no longer observed
-      for (const [atom, sub] of subs) {
-        if (!observedAtoms.has(atom)) {
-          sub.unsubscribe();
-          subs.delete(atom);
-        }
+      return this.currentValue;
+    },
+
+    set(valueOrFn: T | ((prev: T) => T)): void {
+      const compare = options?.compare ?? Object.is;
+      const value =
+        typeof valueOrFn === 'function'
+          ? (valueOrFn as (prev: T) => T)(this.currentValue)
+          : valueOrFn;
+      if (compare(this.currentValue, value)) return;
+      this.currentValue = value;
+      const subs = this.subs;
+      if (subs !== undefined) {
+        propagate(subs);
+        processEffectNotifications();
       }
+    },
 
-      current.value = newValue;
-      observers?.forEach((o) => o.next?.(newValue));
+    subscribe(observerOrFn: Observer<T> | ((value: T) => void)) {
+      const obs = toObserver(observerOrFn);
+      const observed = { current: false };
+      const e = effect(() => {
+        this.get();
+        if (!observed.current) {
+          observed.current = true;
+        } else {
+          obs.next?.(this.currentValue);
+        }
+      });
+
+      return {
+        unsubscribe: () => {
+          e.stop();
+        }
+      };
     }
+  };
 
-    // Initialize computed value
-    current.value = getValue(read);
-  } else {
-    // Handle static value case
-    current.value = valueOrFn;
-  }
+  return atom;
+}
+
+function computed<T>(
+  getter: (read: <U>(atom: Readable<U>) => U) => T,
+  options?: AtomOptions<T>
+): ReadonlyAtom<T> {
+  const computedAtom: ReadonlyAtom<T> = {
+    // Starts as undefined, but will be set by the computed getter (lazy)
+    currentValue: undefined as T,
```

Comment:
given this is a public property - shouldn't `computed` call `update` before returning `computedAtom` to the caller?

Reply:
It's not really supposed to be "public", not sure how to go about this though